### PR TITLE
split level tests into multiple targets

### PR DESCRIPTION
**Pull request by davidsbailey**
## background

https://github.com/code-dot-org/code-dot-org/pull/17133 and at least one of @joshlory 's PRs is blocked because PhantomJS is running out of memory and crashing while running our level tests in CircleCI. Out-of-memory crashes on CircleCI have also been a source of flakiness on the staging branch itself. 

Initial investigation reveals that memory leaks could be due to https://github.com/mochajs/mocha/issues/1991 or https://github.com/ariya/phantomjs/issues/11390 .

## proposed solution

This PR works around this problem by grouping our level tests by level type, based on which subdirectory of levelSolutions they are in. This allows phantomjs and mocha to reset their memory usage so that it doesn't accumulate across all of our level tests. This approach is mentioned [here](https://github.com/karma-runner/karma-phantomjs-launcher/issues/55#issuecomment-103645515).

Caveat: This solution runs the integration tests which are not level tests multiple times, however since these take a total of 6 seconds each time they are run, fixing this does not seem like a priority.

### switch to ChromeHeadless

Because PhantomJS is one possible source of the memory leaks, switching to Chrome is a potential solution. An initial run of our unit tests using ChromeHeadless looks somewhat promising, however there were many test failures in unit tests alone (see https://circleci.com/gh/code-dot-org/code-dot-org/46126 ) so this could take quite some time to implement. Furthermore, a local run of our integration tests consumed 2.2GB and slowed to a crawl after 55 minutes, suggesting headless chrome may not actually fix the problem.

### manual gc

According to https://github.com/mochajs/mocha/issues/1991, Mocha fails to garbage collect references which are initialized in the `before` step and not explicitly set to null in the `after` step. Initial attempts https://github.com/code-dot-org/code-dot-org/pull/17133/commits/cd23618d913c9e6b68312cb514fee35ca4f399b3 and https://github.com/code-dot-org/code-dot-org/pull/17133/commits/9b32ffb74c2f3548387af8f38287e2498e11428b to manually garbage collect a few more refs were not effective.

### load phantomjs images

Many reports on the internet say that telling phantomjs to load images relieves some memory problems, however attempting to do this in https://github.com/code-dot-org/code-dot-org/pull/17133/commits/1e5a5da25619f8925928fb832af3fbb35a81cc3e was not effective.

### CODE REVIEW

File: apps/test-low-memory.sh [hunk 1]
```diff
@@ -11,16 +11,26 @@ if [ -n "$CIRCLECI" ]; then \
...
 fi && \
 
 node --max_old_space_size=4096 `npm bin`/grunt scratchTest && \
-if [-n "$CIRCLECI" ]; then \
+if [ -n "$CIRCLECI" ]; then \
     bash <(curl -s https://codecov.io/bash) -cF scratch; \
 fi && \
 
-node --max_old_space_size=4096 `npm bin`/grunt integrationTest && \
+LEVEL_TYPE='maze|turtle|bounce|calc|eval|flappy|studio' node --max_old_space_size=4096 `npm bin`/grunt integrationTest && \
```
Comment (islemaster) [inline apps/test-low-memory.sh:23]: I'm guessing tests run faster with these grouped, but have we tried doing each of them separately?  I wonder if we'd build much smaller bundles and get some speedup just treating each of these as a different entry point.

Comment (davidsbailey) [reply apps/test-low-memory.sh:23]: I have not. The startup time for each call to `grunt integrationTest` is about 25 seconds on my laptop. I've opened https://github.com/code-dot-org/code-dot-org/pull/17200 to see how this performs on circle, and if it looks good, to merge as a subsequent PR.

File: apps/test/integration/util/testCollectionUtils.js [hunk 2]
```diff
@@ -1,17 +1,43 @@
+/* global LEVEL_TYPE */
+
 var _ = require('lodash');
 
+/**
+ * Given a list of levelSolutions subdirectories, returns only those matching
+ * the pattern specified in LEVEL_TYPE. If specified, LEVEL_TYPE must be contain
+ * one or more subdirectory names joined with '|' e.g. 'applab' or
+ * 'maze|artist|craft'.
```
Comment (Bjvanminnen) [inline apps/test/integration/util/testCollectionUtils.js:9]: I wonder if at some point we should provide a good way to get instructions/information about all of our different test options (i.e. something like `npm run test:help`. At the least, should w ebe documenting LEVEL_TYPE in our readme somewhere?

Comment (davidsbailey) [reply apps/test/integration/util/testCollectionUtils.js:9]: an `npm test:help` would be nice. For now I've added docs for this new env variable in apps/README.md.

**Pull request merged**